File: src/android/plugin/google/maps/experimental/MarkerJsonData.java
package plugin.google.maps.experimental;

import java.util.ArrayList;
import java.util.List;

import org.json.JSONException;
import org.json.JSONObject;

import android.graphics.Point;

import com.google.android.gms.maps.model.LatLng;

public class MarkerJsonData {
  public LatLng position;
  public JSONObject options;
  public Point point;
  private String[] geocells;
  private String markerId;
  
  public MarkerJsonData(JSONObject markerOptions) {
    options = markerOptions;
    JSONObject positionJson;
    List<String> geocellList = new ArrayList<String>();
    try {
      positionJson = markerOptions.getJSONObject("position");
      double lat = positionJson.getDouble("lat");
      double lng = positionJson.getDouble("lng");
      this.position = new LatLng(lat, lng);
      
      for (int i = 0; i < AsyncCluster.MAX_GEOCELL_RESOLUTION; i++) {
        geocellList.add(MarkerClusterUtil.getGeocell(lat, lng, i));
      }
      geocells = geocellList.toArray(new String[]{});
    } catch (JSONException e) {
      e.printStackTrace();
    }
    
    try {
      markerId = markerOptions.getString("id");
    } catch (JSONException e) {
      e.printStackTrace();
    }
  }
  
  public String getGeocell(int zoom) {
    zoom = Math.min(zoom, geocells.length - 1);
    return this.geocells[zoom];
  }
  
  public String getMarkerId() {
    return markerId;
  }
}

File: src/ios/experimental/MarkerJsonData.m
//
//  MarkerJsonData.m
//  DevApp
//
//  Created by Katsumata Masashi on 11/9/14.
//
//

#import "MarkerJsonData.h"

@implementation MarkerJsonData

- (id)initWithOptions:(NSDictionary *) markerOptions {
  if(self == [super init]) {
    self.options = markerOptions;
    
    NSMutableArray *geocellList = [NSMutableArray array];
    
    NSObject *positionJson = [markerOptions valueForKey:@"position"];
    if (positionJson) {
      self.position = CLLocationCoordinate2DMake([[positionJson valueForKey:@"lat"] doubleValue], [[positionJson valueForKey:@"lng"] doubleValue]);
      
      for (int i = 0; i < MAX_GEOCELL_RESOLUTION; i++) {
        [geocellList addObject:[MarkerClusterUtil
                                      getGeocell:self.position.latitude lng:self.position.longitude resolution:i]];
      }
      self.geocells = [geocellList copy];
    }
    
    self.markerId = [markerOptions valueForKey:@"id"];
  }
  return self;
}

- (NSString *)getGeocell:(int) zoom {
  zoom = MIN(zoom, [self.geocells count] - 1);
  return [self.geocells objectAtIndex:zoom];
}

- (NSString *)getMarkerId {
  return self.markerId;
}
@end
